effects/ox/src/main/scala/sttp/client4/ox/ws/OxWebSockets.scala

```diff
@@ -0,0 +1,53 @@
+package sttp.client4.ox.ws
+
+import ox.*
+import ox.channels.Channel
+import ox.channels.ChannelClosed
+import ox.channels.Sink
+import ox.channels.Source
+import sttp.client4.ws.SyncWebSocket
+import sttp.ws.WebSocketFrame
+
+import scala.util.control.NonFatal
+
+  extension (ws: SyncWebSocket)(using Ox) def asSourceAndSink: (Source[WebSocketFrame], Sink[WebSocketFrame]) = 
+    val srcChannel = Channel.bufferedDefault[WebSocketFrame]
+    fork {
+      repeatWhile {
+        try ws.receive() match 
+          case WebSocketFrame.Close(status, _) if status > 1001 => 
+              srcChannel.error(new Exception(s"WebSocket closed with status $status"))
+              false
+          case _: WebSocketFrame.Close => 
+              srcChannel.done()
+              false
+          case frame =>
+              srcChannel.send(frame)
+              true
+        catch case NonFatal(err) => 
+          srcChannel.error(err)          
+          false
+        }
+    }
+    val sinkChannel = Channel.bufferedDefault[WebSocketFrame]
+    fork {
+      try 
+        repeatWhile {
+          sinkChannel.receiveOrClosed() match
+            case closeFrame: WebSocketFrame.Close => 
+                ws.send(closeFrame) // TODO should we just let 'send' throw exceptions?
+                false
+            case frame: WebSocketFrame =>
+                ws.send(frame)
+                true
+            case ChannelClosed.Done =>               
+                ws.send(WebSocketFrame.close)
+                false
+            case ChannelClosed.Error(err) =>
+                // There's no proper "client error" status. Statuses 4000+ are available for custom cases
+                ws.send(WebSocketFrame.Close(4000, "Client error")) // TODO should we bother the server with client error?
```

Comment:
shouldn't we also then complete the `srcChannel`?

Reply:
This question made me rework how source and sink interact. I implemented following logic:
1. When processing responses fails, close the request channel
2. When processing requests fails, close the response channel
3. When processing responses completes (a `Close` frame is received), close the request channel

There's one more thing that I'm not sure about: how to handle a completed request channel - when `channel.receiveOrClosed()` returns a `ChannelClosed.Done`. Current logic handles this by sending a `Close` frame, but I'm not sure we want it. For example, sending requests like:
```scala
    someSource.map(i => WebSocketFrame.text(s"test$i")).pipeTo(ws)
```
Sends a bunch of frames *immediately* followed by a `Close` frame when the source ends. What should we expect from the backend in such a case? I compared tapir-http4s with tapir-netty-sync:
* `tapir-http4s` processes the `Close` frame sequentially after processing all the previous frames
* `tapir-netty-sync` handles `Close` when it arrives, and sends a `Close` back immediately, regardless of all the frames that are "in-flight".
There is no specification for how a WS endpoint should behave, and I'm not sure myself which mode is better as a default.

Reply:
I added an issue for tapir-netty servers https://github.com/softwaremill/tapir/issues/3776